File: app/src/main/java/com/cesoft/cesrssreader/net/RssNet.java
package com.cesoft.cesrssreader.net;

import android.util.Log;

import com.cesoft.cesrssreader.Util;
import com.cesoft.cesrssreader.model.RssFeedModel;
import com.cesoft.cesrssreader.model.RssItemModel;
import com.cesoft.cesrssreader.model.RssSourceModel;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;

////////////////////////////////////////////////////////////////////////////////////////////////////
// Created by Cesar Casanova on 21/06/2017.
public class RssNet
{
	private static final String TAG = RssNet.class.getSimpleName();
	//private String _link;
	//public RssNet(String link) { _link = link; }

	////////////////////////////////////////////////////////////////////////////////////////////////
	public interface Callback
	{
		void onData(RssFeedModel data);
		void onError(Throwable err);
	}

	//----------------------------------------------------------------------------------------------
	public static void fetch(String link, final Callback cb)//"http://www.feedforall.com/sample.xml"
	{
		if(link == null || link.isEmpty())
		{
			Log.e(TAG, "fetch:e: Rss Feed Url is wrong : "+link);
			cb.onError(new Exception("fetch:e: Rss Feed Url is wrong : "+link));
			return;
		}
		//final List<RssModel> items_ = new ArrayList<>();
		//RssNet reader = new RssNet();
		RssNet.getItems(link)
				.subscribeOn(Schedulers.newThread())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(new Action1<List<RssFeedModel>>()
				{
					@Override
					public void call(List<RssFeedModel> items)
					{
						cb.onData(items.get(0));
						//items_.addAll(items);
					}
				},
				new Action1<Throwable>()
				{
					@Override
					public void call(Throwable e)
					{
						cb.onError(e);
						Log.e("App", "Failed to download RSS Items:"+e);
					}
				});
	}


	//----------------------------------------------------------------------------------------------
	private static Observable<List<RssFeedModel>> getItems(final String link)
	{
		return Observable.create(new Observable.OnSubscribe<List<RssFeedModel>>()
		{
			@Override
			public void call(Subscriber<? super List<RssFeedModel>> subscriber)
			{
				try
				{
					SAXParserFactory factory = SAXParserFactory.newInstance();
					SAXParser saxParser = factory.newSAXParser();
					//Creates a new RssHandler which will do all the parsing.
					RssHandler handler = new RssHandler();
					//Pass SaxParser the RssHandler that was created.
					saxParser.parse(link, handler);
					RssFeedModel rssFeed = handler.getRssFeed(link);
					ArrayList<RssFeedModel> list = new ArrayList<>();
					list.add(rssFeed);
					subscriber.onNext(list);
					subscriber.onCompleted();
				}
				catch(Exception e)
				{
					subscriber.onError(e);
				}
			}
		});
	}

	////////////////////////////////////////////////////////////////////////////////////////////////
	private static class RssHandler extends DefaultHandler
	{
		private List<RssItemModel> rssItemList;
		private RssItemModel rssItem;
		private RssSourceModel rssSource;
		private boolean parsingTitle;
		private boolean parsingLink;
		private boolean parsingDescription;
		private boolean parsingFecha;
		private String fecha;

		//------------------------------------------------------------------------------------------
		RssHandler()
		{
			rssItemList = new ArrayList<>();
			rssSource = new RssSourceModel();
		}
		//------------------------------------------------------------------------------------------
		RssFeedModel getRssFeed(String link)
		{
			rssSource.setLink(link);
			return new RssFeedModel(rssSource, rssItemList);
		}

		//------------------------------------------------------------------------------------------
		//Called when an opening tag is reached, such as <item> or <title>
		@Override
		public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException
		{
			//Log.e(TAG, "startElement: ---------------------------"+rssItemList.size()+"-----------------------------"+qName);
			switch(qName.toLowerCase())
			{
			case "item":
				rssItem = new RssItemModel();
				fecha = "";
				break;
			case "title":
				parsingTitle = true;
				break;
			case "link":
				parsingLink = true;
				break;
			case "description":
				parsingDescription = true;
				break;
			case "pubdate":
				parsingFecha = true;
				break;
			case "enclosure":
				if(attributes.getValue("type") == null || !attributes.getValue("type").contains("image"))break;
			case "media:thumbnail":
			case "media:content":
			case "image":
				if(attributes.getValue("url") != null)
				{
					if(rssItem != null)
						rssItem.setImg(attributes.getValue("url"));
					else
						rssSource.setImg(attributes.getValue("url"));
				}
				break;
			}
		}

		//------------------------------------------------------------------------------------------
		//Called when a closing tag is reached, such as </item> or </title>
		@Override
		public void endElement(String uri, String localName, String qName) throws SAXException
		{
			switch(qName.toLowerCase())
			{
			case "item":
				//End of an item so add the currentItem to the list of items.
				rssItemList.add(rssItem);
				rssItem = null;
				break;
			case "title":
				parsingTitle = false;
				break;
			case "link":
				parsingLink = false;
				break;
			case "description":
				parsingDescription = false;
				break;
			case "pubdate":
				parsingFecha = false;
				fecha = "";
				break;
			}
		}

		//------------------------------------------------------------------------------------------
		//Goes through character by character when parsing whats inside of a tag.
		@Override
		public void characters(char[] ch, int start, int length) throws SAXException
		{
			if(rssItem != null)
			{
				//If parsingTitle is true, then that means we are inside a <title> tag so the text is the title of an item.
				if(parsingTitle)
				{
					String s = "";
					if(rssItem.getTitulo() != null)s=rssItem.getTitulo();
					rssItem.setTitulo(s + new String(ch, start, length));
					Log.e(TAG, "TITULO: ------------- "+new String(ch, start, length));
				}
				else if(parsingDescription)
				{
					String s = "";
					if(rssItem.getDescripcion() != null)s=rssItem.getDescripcion();
					rssItem.setDescripcion(s + new String(ch, start, length));
				}
				else if(parsingLink)
				{
					String s = "";
					if(rssItem.getLink() != null)s=rssItem.getLink();
					rssItem.setLink(s + new String(ch, start, length));
				}
				else if(parsingFecha)
				{
					fecha += new String(ch, start, length);
					rssItem.setFecha(Util.str2date(fecha));
				}
			}
			else
			{
				if(parsingTitle)
					rssSource.setTitulo(new String(ch, start, length));
				else if(parsingDescription)
					rssSource.setDescripcion(new String(ch, start, length));
				else if(parsingFecha)
					rssSource.setFecha(Util.str2date(new String(ch, start, length)));
				else if(parsingLink)
					Log.e(TAG, "-------- FEED LINK : "+new String(ch, start, length));
				//rssSource.setLink(new String(ch, start, length));
			}
		}
	}
}

File: app/src/main/java/com/cesoft/cesrssreader/net/RssNet.kt
package com.cesoft.cesrssreader.net

import com.cesoft.cesrssreader.Util
import com.cesoft.cesrssreader.model.RssFeedModel
import com.cesoft.cesrssreader.model.RssItemModel
import com.cesoft.cesrssreader.model.RssSourceModel
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

import org.xml.sax.Attributes
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler

import java.util.ArrayList

import javax.xml.parsers.SAXParserFactory



////////////////////////////////////////////////////////////////////////////////////////////////////
// Created by Cesar Casanova on 21/06/2017.
class RssNet
{
	////////////////////////////////////////////////////////////////////////////////////////////////
	interface Callback
	{
		fun onData(data: RssFeedModel)
		fun onError(err: Throwable)
	}

	////////////////////////////////////////////////////////////////////////////////////////////////
	companion object
	{
		val TAG = RssNet::class.java.simpleName

		//----------------------------------------------------------------------------------------------
		fun fetch(link: String?, cb: Callback)//"http://www.feedforall.com/sample.xml"
		{
			if(link == null || link.isEmpty())
			{
				Util.log(TAG, "fetch:e: Rss Feed Url is wrong : " + link!!)
				cb.onError(Exception("fetch:e: Rss Feed Url is wrong : " + link))
				return
			}
			//final List<RssModel> items_ = new ArrayList<>();
			//RssNet reader = new RssNet();
			getItems(link)
				.subscribeOn(Schedulers.newThread())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(
						{
							items : List<RssFeedModel> -> cb.onData(items[0])
						},
						{
							e : Throwable ->
								cb.onError(e)
								Util.log("App", "Failed to download RSS Items:" + e)
						})
		}

		//----------------------------------------------------------------------------------------------
		private fun getItems(link: String): Observable<List<RssFeedModel>>
		{
			return Observable.create { subscriber ->
				try
				{
					val factory = SAXParserFactory.newInstance()
					val saxParser = factory.newSAXParser()
					//Creates a new RssHandler which will do all the parsing.
					val handler = RssHandler()
					//Pass SaxParser the RssHandler that was created.
					saxParser.parse(link, handler)
					val rssFeed = handler.getRssFeed(link)
					val list = ArrayList<RssFeedModel>()
					list.add(rssFeed)
					subscriber.onNext(list)
					subscriber.onComplete()// .onCompleted()
				}
				catch(e: Exception)
				{
					subscriber.onError(e)
				}
			}
		}
	}



	////////////////////////////////////////////////////////////////////////////////////////////////
	class RssHandler : DefaultHandler()
	{
		private var rssItemList: MutableList<RssItemModel> = ArrayList()
		private var rssItem: RssItemModel? = null
		private var rssSource: RssSourceModel = RssSourceModel()
		private var parsingTitle: Boolean = false
		private var parsingLink: Boolean = false
		private var parsingDescription: Boolean = false
		private var parsingFecha: Boolean = false
		private var fecha: String? = null


		//------------------------------------------------------------------------------------------
		internal fun getRssFeed(link: String): RssFeedModel
		{
			rssSource.link = link
			return RssFeedModel(rssSource, rssItemList)
		}

		//------------------------------------------------------------------------------------------
		//Called when an opening tag is reached, such as <item> or <title>
		@Throws(SAXException::class)
		override fun startElement(uri: String, localName: String, qName: String, attributes: Attributes)
		{
			//Log.e(TAG, "startElement: ---------------------------"+rssItemList.size()+"-----------------------------"+qName);
			when(qName.toLowerCase())
			{
				"item" ->
				{
					rssItem = RssItemModel()
					fecha = ""
				}
				"title" -> parsingTitle = true
				"link" -> parsingLink = true
				"description" -> parsingDescription = true
				"pubdate" -> parsingFecha = true
				"enclosure" ->
				{
					if(attributes.getValue("type") != null
						&& attributes.getValue("type")!!.contains("image")
						&& attributes.getValue("url") != null)
					{
						if(rssItem != null)
							rssItem!!.img = attributes.getValue("url")
						else
							rssSource.img = attributes.getValue("url")
					}
				}
				"media:thumbnail", "media:content", "image" -> if(attributes.getValue("url") != null)
				{
					if(rssItem != null)
						rssItem!!.img = attributes.getValue("url")
					else
						rssSource.img = attributes.getValue("url")
				}
			}
		}

		//------------------------------------------------------------------------------------------
		//Called when a closing tag is reached, such as </item> or </title>
		@Throws(SAXException::class)
		override fun endElement(uri: String, localName: String, qName: String)
		{
			when(qName.toLowerCase())
			{
				"item" -> {
					//End of an item so add the currentItem to the list of items.
					//if(rssItem!=null)
						rssItemList.add(rssItem!!)
					rssItem = null
				}
				"title" -> parsingTitle = false
				"link" -> parsingLink = false
				"description" -> parsingDescription = false
				"pubdate" -> {
					parsingFecha = false
					fecha = ""
				}
			}
		}

		//------------------------------------------------------------------------------------------
		//Goes through character by character when parsing whats inside of a tag.
		@Throws(SAXException::class)
		override fun characters(ch: CharArray, start: Int, length: Int)
		{
			if(rssItem != null)
			{
				//If parsingTitle is true, then that means we are inside a <title> tag so the text is the title of an item.
				if(parsingTitle)
				{
					var s = ""
					if(rssItem!!.titulo != null) s = rssItem!!.titulo
					rssItem!!.titulo = s + String(ch, start, length)
					Util.log(TAG, "TITULO: ------------- " + String(ch, start, length))
				}
				else if(parsingDescription)
				{
					var s = ""
					if(rssItem!!.descripcion != null) s = rssItem!!.descripcion
					rssItem!!.descripcion = s + String(ch, start, length)
				}
				else if(parsingLink)
				{
					var s = ""
					if(rssItem!!.link != null) s = rssItem!!.link
					rssItem!!.link = s + String(ch, start, length)
				}
				else if(parsingFecha)
				{
					fecha += String(ch, start, length)
					rssItem!!.fecha = Util.str2date(fecha)
				}
			}
			else
			{
				when
				{
					parsingTitle -> rssSource.titulo = String(ch, start, length)
					parsingDescription -> rssSource.descripcion = String(ch, start, length)
					parsingFecha -> rssSource.fecha = Util.str2date(String(ch, start, length))
					parsingLink -> Util.log(TAG, "-------- FEED LINK : " + String(ch, start, length))
				}
			}
		}
	}
}